File: implementations/java/src/main/java/io/rcl/Converters.java
package io.rcl;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class Converters {
  static Map<String, Object> toObject(DocumentNode doc) {
    if (doc.rootValue != null) {
      Map<String, Object> root = new LinkedHashMap<>();
      root.put("root", val(doc.rootValue));
      return root;
    }
    Map<String, Object> out = new LinkedHashMap<>();
    for (BlockNode b : doc.blocks) {
      if (b.argument == null) out.put(b.name, blockObj(b));
      else {
        Map<String, Object> x = new LinkedHashMap<>();
        x.put(b.argument, blockObj(b));
        out.put(base(b.name), x);
      }
    }
    return out;
  }

  private static Map<String, Object> blockObj(BlockNode b) {
    Map<String, Object> out = new LinkedHashMap<>();
    for (String k : b.propertyOrder) putPath(out, k, val(b.properties.get(k)));
    for (BlockNode c : b.blocks) mergeAt(out, c.name, blockObj(c));
    for (BlockNode c : b.namedBlocks) {
      String bn = base(c.name);
      Map<String, Object> m = castMap(out.getOrDefault(bn, new LinkedHashMap<>()));
      m.put(c.argument, blockObj(c));
      out.put(bn, m);
    }
    return out;
  }

  private static void putPath(Map<String, Object> out, String k, Object v) {
    String[] p = k.split("\\.");
    Map<String, Object> cur = out;
    for (int i = 0; i < p.length - 1; i++) cur = castMap(cur.computeIfAbsent(p[i], z -> new LinkedHashMap<>()));
    cur.put(p[p.length - 1], v);
  }

  private static void mergeAt(Map<String, Object> out, String k, Map<String, Object> v) {
    Map<String, Object> c = castMap(out.getOrDefault(k, new LinkedHashMap<>()));
    c.putAll(v);
    out.put(k, c);
  }

  private static Object val(AstNode n) {
    if (n instanceof StringNode s) return s.value;
    if (n instanceof NumberNode d) return d.value;
    if (n instanceof BooleanNode b) return b.value;
    if (n instanceof BlockNode b) return blockObj(b);
    List<Object> a = new ArrayList<>();
    for (AstNode x : ((ArrayNode) n).elements) a.add(val(x));
    return a;
  }

  @SuppressWarnings("unchecked")
  private static Map<String, Object> castMap(Object x) { return (Map<String, Object>) x; }
  private static String base(String n) { return n.endsWith("s") ? n : n + "s"; }

  static String toYAML(DocumentNode doc) { return yaml(toObject(doc), 0); }
  static String toTOML(DocumentNode doc) { StringBuilder o = new StringBuilder(); toml(o, toObject(doc), ""); return o.toString(); }
  static String toHCL(DocumentNode doc) { return hcl(toObject(doc), 0); }

  private static String yaml(Map<String, Object> m, int n) {
    StringBuilder o = new StringBuilder();
    for (var e : m.entrySet()) {
      String p = " ".repeat(n * 2);
      if (e.getValue() instanceof Map<?, ?> mm) o.append(p).append(e.getKey()).append(":\n").append(yaml(castMap(mm), n + 1));
      else o.append(p).append(e.getKey()).append(": ").append(scalar(e.getValue())).append("\n");
    }
    return o.toString();
  }

  private static void toml(StringBuilder o, Map<String, Object> m, String p) {
    if (!p.isEmpty()) o.append("[").append(p).append("]\n");
    for (var e : m.entrySet()) if (!(e.getValue() instanceof Map<?, ?>)) o.append(e.getKey()).append(" = ").append(scalar(e.getValue())).append("\n");
    for (var e : m.entrySet()) if (e.getValue() instanceof Map<?, ?> mm) {
      o.append("\n");
      toml(o, castMap(mm), p.isEmpty() ? e.getKey() : p + "." + e.getKey());
    }
  }

  private static String hcl(Map<String, Object> m, int n) {
    StringBuilder o = new StringBuilder();
    for (var e : m.entrySet()) {
      String p = " ".repeat(n * 2);
      if (e.getValue() instanceof Map<?, ?> mm) o.append(p).append(e.getKey()).append(" {\n").append(hcl(castMap(mm), n + 1)).append(p).append("}\n");
      else o.append(p).append(e.getKey()).append(" = ").append(scalar(e.getValue())).append("\n");
    }
    return o.toString();
  }

  private static String scalar(Object v) {
    if (v instanceof String s) return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\t", "\\t") + "\"";
    if (v instanceof Boolean b) return b ? "true" : "false";
    if (v instanceof List<?> a) {
      StringBuilder o = new StringBuilder("[");
      for (int i = 0; i < a.size(); i++) { if (i > 0) o.append(", "); o.append(scalar(a.get(i))); }
      return o.append("]").toString();
    }
    return String.valueOf(v);
  }
}

File: implementations/kotlin/src/main/kotlin/io/rcl/Converters.kt
package io.rcl

object Converters {
  fun toObject(document: DocumentNode): Map<String, Any?> {
    if (document.blocks.isEmpty()) return emptyMap()
    val out = linkedMapOf<String, Any?>()
    document.blocks.forEach { root ->
      if (root.argument != null) out[namedBase(root.name)] = mapOf(root.argument!! to blockToMap(root))
      else out[root.name] = blockToMap(root)
    }
    return out
  }

  fun toYaml(document: DocumentNode): String = emitYaml(toObject(document), 0)
  fun toToml(document: DocumentNode): String = emitToml(toObject(document))
  fun toHcl(document: DocumentNode): String = emitHcl(toObject(document), 0)

  private fun blockToMap(block: BlockNode): MutableMap<String, Any?> {
    val result = linkedMapOf<String, Any?>()
    block.properties.forEach { (k, v) -> insertPath(result, k, nodeToAny(v)) }

    val children = uniqueChildren(block)
    children.filter { it.argument == null }.forEach { child ->
      val childMap = blockToMap(child)
      val existing = result[child.name] as? Map<String, Any?>
      if (existing != null) childMap.putAll(existing)
      result[child.name] = childMap
    }
    children.filter { it.argument != null }.forEach { child ->
      val base = namedBase(child.name)
      val parent = (result[base] as? MutableMap<String, Any?>) ?: linkedMapOf()
      parent[child.argument!!] = blockToMap(child)
      result[base] = parent
    }
    return result
  }

  private fun uniqueChildren(block: BlockNode): List<BlockNode> {
    val out = mutableListOf<BlockNode>()
    val seen = mutableSetOf<String>()
    val fromMap = block.blocks.values.toList().sortedBy { (it.argument?.let { a -> "${it.name}:$a" } ?: it.name) }
    (fromMap + block.namedBlocks).forEach { child ->
      val key = child.argument?.let { a -> "${child.name}:$a" } ?: child.name
      if (seen.add(key)) out += child
    }
    return out
  }

  private fun nodeToAny(node: AstNode): Any? = when (node) {
    is StringNode -> node.value
    is NumberNode -> node.value
    is BooleanNode -> node.value
    is ArrayNode -> node.elements.map { nodeToAny(it) }
    else -> null
  }

  private fun emitYaml(v: Any?, indent: Int): String {
    val pad = "  ".repeat(indent)
    return when (v) {
      is Map<*, *> -> v.keys.map { it.toString() }.sorted().joinToString("\n") { k ->
        val item = v[k]
        if (item is Map<*, *> || item is List<*>) "$pad$k:\n${emitYaml(item, indent + 1)}"
        else "$pad$k: ${scalar(item)}"
      }
      is List<*> -> v.joinToString("\n") { item ->
        if (item is Map<*, *> || item is List<*>) "$pad-\n${emitYaml(item, indent + 1)}"
        else "$pad- ${scalar(item)}"
      }
      else -> "$pad${scalar(v)}"
    }
  }

  private fun emitToml(root: Map<String, Any?>): String {
    val out = mutableListOf<String>()
    fun walk(obj: Map<String, Any?>, prefix: String?) {
      obj.keys.sorted().forEach { k -> if (obj[k] !is Map<*, *>) out += "$k = ${scalar(obj[k])}" }
      obj.keys.sorted().forEach { k ->
        val child = obj[k] as? Map<String, Any?> ?: return@forEach
        val sec = if (prefix == null) k else "$prefix.$k"
        if (out.isNotEmpty()) out += ""
        out += "[$sec]"
        walk(child, sec)
      }
    }
    walk(root, null)
    return out.joinToString("\n")
  }

  private fun emitHcl(v: Any?, indent: Int): String {
    val pad = "  ".repeat(indent)
    val obj = v as? Map<String, Any?> ?: return "$pad${scalar(v)}"
    return obj.keys.sorted().joinToString("\n") { k ->
      val item = obj[k]
      if (item is Map<*, *>) "$pad$k {\n${emitHcl(item, indent + 1)}\n$pad}"
      else "$pad$k = ${scalar(item)}"
    }
  }

  private fun scalar(v: Any?): String = when (v) {
    is String -> "\"${v.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\t", "\\t")}\""
    is Boolean -> if (v) "true" else "false"
    is Number -> v.toString()
    is List<*> -> "[${v.joinToString(", ") { scalar(it) }}]"
    else -> "{}"
  }

  private fun insertPath(target: MutableMap<String, Any?>, key: String, value: Any?) {
    val parts = key.split(".")
    if (parts.size == 1) {
      if (target.containsKey(key)) throw IllegalArgumentException("duplicate key")
      target[key] = value
      return
    }
    val head = parts.first()
    val existing = target[head]
    if (existing != null && existing !is MutableMap<*, *> && existing !is Map<*, *>) {
      throw IllegalArgumentException("key conflict")
    }
    val branch = (existing as? MutableMap<String, Any?>) ?: (existing as? Map<String, Any?>)?.toMutableMap() ?: linkedMapOf()
    insertPath(branch, parts.drop(1).joinToString("."), value)
    target[head] = branch
  }

  private fun namedBase(name: String): String = if (name.endsWith("s")) name else "${name}s"
}
